lib/scholar/cluster/hierarchical.ex

```diff
@@ -0,0 +1,476 @@
+          pointers = Nx.indexed_put(pointers, indices, Nx.stack([c, c]))
+
+          {{clades, count + 1, pointers, pairwise, diss, sizes}, links}
+        end
+      end
+
+    {clades, count, pointers, pairwise, diss, sizes}
+  end
+
+  defnp find_clade(pointers, i) do
+    {i, _, _} =
+      while {_current = i, next = pointers[i], pointers}, next != -1 do
+        {next, pointers[next], pointers}
+      end
+
+    i
+  end
+
+  # Dissimilarity functions
+
+  defnp pairwise_euclidean(%Nx.Tensor{} = x) do
+    x |> pairwise_euclidean_sq() |> Nx.sqrt()
+  end
+
+  defnp pairwise_euclidean_sq(%Nx.Tensor{} = x) do
+    sq = Nx.sum(x ** 2, axes: [1], keep_axes: true)
+    sq + Nx.transpose(sq) - 2 * Nx.dot(x, [1], x, [1])
+  end
+
+  # Dissimilarity update functions
+
+  defnp average(dac, dbc, _dab, sa, sb, _sc),
+    do: (sa * dac + sb * dbc) / (sa + sb)
+
+  # defnp centroid(dac, dbc, dab, sa, sb, _sc),
+  #   do: Nx.sqrt((sa * dac + sb * dbc) / (sa + sb) - sa * sb * dab / (sa + sb) ** 2)
+
+  defnp complete(dac, dbc, _dab, _sa, _sb, _sc),
+    do: Nx.max(dac, dbc)
+
+  # defnp median(dac, dbc, dab, _sa, _sb, _sc),
+  #   do: Nx.sqrt(dac / 2 + dbc / 2 - dab / 4)
+
+  defnp single(dac, dbc, _dab, _sa, _sb, _sc),
+    do: Nx.min(dac, dbc)
+
+  defnp ward(dac, dbc, dab, sa, sb, sc),
+    do: Nx.sqrt(((sa + sc) * dac + (sb + sc) * dbc - sc * dab) / (sa + sb + sc))
+
+  defnp weighted(dac, dbc, _dab, _sa, _sb, _sc),
+    do: (dac + dbc) / 2
+
+  # Predict functions
+
+  @predict_opts_schema [
+    cluster_by: [
+      type: :non_empty_keyword_list,
+      required: true,
+      keys: [
+        height: [
+          type: :float,
+          doc: "Height of the dendrogram to use as the split point for clusters."
+        ],
+        num_clusters: [
+          type: :pos_integer,
+          doc: "Number of clusters to form."
+        ]
+      ],
+      doc: """
+      How to select which clades from the dendrogram should form the final clusters.
+      Must provide either a height or a number of clusters.
+      """
+    ]
+  ]
+  @doc """
+  Cluster the dataset using hierarchical clustering or cluster the model built by `fit/2`.
+
+  ## Options
+
+  If the input is a `Scholar.Cluster.Hierarchical` struct, only these options are required:
+
+  #{NimbleOptions.docs(@predict_opts_schema)}
+
+  If the input is a `Nx.Tensor`, you must also pass the options required by `fit/2`.
+
+  ## Return values
+
+  Returns a `Nx.Tensor` with shape `{model.num_points}` and values `0..(k - 1)` where `k` is the
+  number of clusters formed.
+  The `i`th element of the result tensor is the label of the `i`th data point's cluster.
+  (Cluster labels are arbitrary, but deterministic.)
+
+  ## Examples
+
+      iex> data = Nx.tensor([[2], [7], [9], [0], [3]])
+      iex> model = Hierarchical.fit(data)
+      iex> Hierarchical.fit_predict(model, cluster_by: [num_clusters: 3])
+      Nx.tensor([0, 1, 1, 2, 0])
+  """
+  deftransform fit_predict(data_or_model, opts \\ [])
```

Comment:
This isn't really a `deftransform` function. They are all assuming we can access the values in the tensor and `defn` and `deftransform` cannot access values, only types and shapes. My suggestion is to call this function `labels` or similar, no `deftransform` or `defn`, and document that their goal is to return a high-level Elixir data structure representing the hierarchies (it can even be a map, not a tensor, because this is not working inside `defn` anyway).

That's the key philosophy inside Scholar/Nx: everything that can be done inside Nx, is explicitly done inside Nx. If you can't do it inside Nx, then that's fine too, but then it should be separated and marked as such!

Reply:
The philosophy is important for the GPU, because going in and out is often expensive, especially when we consider immutability. We can only be immutable inside `defn`.

Reply:
Awesome thanks for making those points! I'm definitely still cargo-culting much of the nuance.

Alright, what do we want the API to be? I tried to use `affinity_propagation` as my guide, but it's not a perfect fit.

We need `fit` to return a `%Hierarchical{}` struct so we can theoretically plot the dendrogram before we choose how to cluster. But other than that, there are a bunch of ways to organize things.

We can do basically what we have now:

* `deftransform fit` : `data: %Tensor{}` -> `model: %Hierarchical{}`
* `def label` : `model: %Hierarchical{}` -> `labels: %Tensor{}`

where `labels` is what `fit_predict` returns now.

We can make `labels` field of the `%Hierarchical{}` struct. Then `label` just sets that field and returns the altered `%Hierarchical{}`.

We can have `label` take a `return_type` option. Then you could either have it return a tensor of labels or a map of `%{cluster_label => list_of_member_indices}`.

We can have label return a `%HierarchicalClusters{}` struct that has both a `labels` field and a `clusters` field which would hold the map from the previous option.

I think all of those are reasonable options (and there are more). I'm happy to pick one! But I think you all likely know better what fits best with the ecosystem.

Reply:
> deftransform fit : data: %Tensor{} -> model: %Hierarchical{}

Agreed.

And then let's add `labels_map` that returns `%{cluster_label => list_of_member_indices}`?